src/nvim/ex_getln.c

```diff
@@ -2181,19 +2207,18 @@ static int command_line_handle_key(CommandLineState *s)
     return command_line_changed(s);
   }
 
-  // put the character in the command line
-  if (IS_SPECIAL(s->c) || mod_mask != 0) {
-    put_on_cmdline(get_special_key_name(s->c, mod_mask), -1, true);
-  } else {
-    int j = utf_char2bytes(s->c, IObuff);
-    IObuff[j] = NUL;                // exclude composing chars
-    put_on_cmdline(IObuff, j, true);
-  }
-  return command_line_changed(s);
+  put_int_on_cmdline(s);
+  return ccline.one_key ? 0 : command_line_changed(s);
 }
 
 static int command_line_not_changed(CommandLineState *s)
 {
+  // For one key prompt, put onto cmdline and exit.
+  if (ccline.one_key) {
+    put_int_on_cmdline(s);
+    return 0;
+  }
+
```

Comment:
Doing this in `command_line_not_changed()` is a bit strange, as the effect of a control character in cmdline mode still takes effect before this is reached. For Ctrl-E it's not very obvious as it just moves the cursor, but the behavior may very strange if one presses some other keys (e.g Ctrl-R).

Reply:
Agreed, I thought some of the logic in `command_line_handle_key()` could still be useful/necessary, but we shouldn't do any of it for one key prompts. (Except `Esc` and `Ctrl_C`, to avoid putting their control codes onto the cmdline.)